lib/storage/metadata/mongoclient/MongoClientInterface.js

```diff
@@ -1162,30 +1172,62 @@ class MongoClientInterface {
         const masterKey = formatMasterKey(objName, params.vFormat);
         const versionKey = formatVersionKey(objName, params.versionId, params.vFormat);
         const _vid = generateVersionId(this.replicationGroupId);
-        const mst = generatePHDVersion(_vid);
-        c.bulkWrite([{
-            updateOne: {
-                filter: {
+        async.waterfall([
+            next => c.findOne({
+                // We try to get both the master object and the latest version
+                // of the object as there are cases where only one of them is available.
+                // When the last version is a delete marker only the version is available.
+                // In versioning suspended buckets, we can have masters that do not have versions,
+                // the oher case is also possible where an object has both the master and versions.
+                // That is why we use the same deletion logic for both versioned and versioning suspended
+                // objects.
+                $or: [{
                     _id: masterKey,
-                },
-                update: {
-                    $set: { _id: masterKey, value: mst },
-                },
-                upsert: true,
-            },
-        }, {
-            deleteOne: {
-                filter: {
+                }, {
                     _id: versionKey,
+                }],
+            }, {}, (err, doc) => {
+                if (err) {
+                    log.error('deleteObjectVerMaster: error getting object',
+                        { bucket: bucketName, object: objName, error: err.message });
+                    return next(errors.InternalError);
+                }
+                if (!doc) {
+                    log.error('deleteObjectVerMaster: unable to find target object',
+                        { bucket: bucketName, object: objName });
+                    return next(errors.NoSuchKey);
+                }
+                const objMetadata = new ObjectMD(doc.value);
+                // adding PHD flag to object metadata
+                objMetadata.setIsPHD(true);
+                // updating versionId
+                objMetadata.setVersionId(_vid);
+                return next(null, objMetadata.getValue());
+            }),
+            // Updating/creating master object
+            (mst, next) => c.updateOne(
+                { _id: masterKey },
+                {
+                    $set: { _id: masterKey, value: mst },
```

Comment:
Somehow, i am a bit inconfortable with this : we get the (mutable) metadata and store in the master, which may create a race condition (e.g. we put an older copy of the metadata).

For many fields this may be acceptable, as they are actually immutable; and for some other fields we don't really care... But putting everything in there means we may later be tempted to use the whole information, even if parts of it may not be correct.

Thus I would like others opinion on this.

Reply:
This master value seems to be either deleted or overwritten by deleteOrRepairPHD or external "PUT/DELETE" object calls.
It seems that, here, we use it exclusively for the MongoDB oplog's delete events https://github.com/scality/Arsenal/pull/1944/files#diff-9c207befbd8f4d79a3b6e3d7d143444e804ca1684236b8fd9fa0f9e03c53477aR1124 and since any "outdated master" won't be deleted, oplog state should be fine.
However, even if it does not seem to be mandatory in our case, we should find a way to avoid potential race conditions (with findAndModify or condition expressions...).

Reply:
To remove any race condition, i decided to stick with the original implementation of the function, where we directly update the master, this should hopefully remove any inconsistency in the metadata.

The original reason of why we first `Find` and then `Update` the metadata is because of one particular case where we don't have a master object. This happens when the last version is a delete marker, and only happens when the bucket is versioned. In this case we don't really have a choice i think but to get the version and then create the master with its data.

So now in this case we'll have an object with empty metadata kind of like it was before, only containing the `isPHD` flag and the versionId, this is not really an issue as the solution to this would just be to use version object's events when dealing with a versioned bucket as they contain everything.